### ISSUE TITLE
[Question] - How to share this CPS functionality

### ISSUE THREAD

**Issue by dazinator** [Microsoft/VSProjectSystem#145]
I have put together a `single file generator` that parses the old webforms style files (`aspx, ascx`) and generates the corresponding designer.cs files for them.

At the moment, I have my own project system that sets this as the `generator` property on aspx and ascx file types when they are added to the project.

However I am wondering how I can tie my functionality in with other CPS project systems that might deal with ascx, and aspx files and require this functionality. Shouldn't my extension be able to check the capability of the host project, and then set the `Generator` property automatically on ascx, aspx files within that project? But then again, what if the host project doesn't want my extension to do this for particular files - can the host project opt out somehow? What I am really trying to flesh out is how can some other project system (not mine) advertise a "web forms" capability, so that my CPS / extension knows to hook into the project, and add the `Generator` for `ascx, asmx` files.

**Comment by dazinator** [comment]
Perhaps I am thinking about this wrong. Perhaps I want to detect projects that **are not** capable of building `aspx`, `ascx` files already, and  where that capability is missing, I want to add my generator to all `aspx`, `ascx` files to cater for that missing capabilty. Does that sound right?

**Comment by adrianvmsft** [comment]
That is a very good question. I think there are a few options:

1. NuGet package that inserts a target in the project that contains that particular capability; this way your customers could enable/disable that feature by adding a NuGet package to their project. This should be possible thanks to a new feature we added in VS15: [Dynamic capabilities](https://github.com/Microsoft/VSProjectSystem/blob/master/doc/overview/dynamicCapabilities.md)
Last time I checked though there were some issues with NuGet support for CPS projects, but we are hoping to address them
2. Custom capabilities provider - that activates based on existing known capabilities, that inserts your capability (eventually based on some options) - see [IProjectCapabilitiesProvider](https://github.com/Microsoft/VSProjectSystem/blob/master/doc/extensibility/IProjectCapabilitiesProvider.md)
3. Activating your functionality based on a known capability - however this could also have side effects, and there would be no way to turn it off besides uninstalling your extension
4. Of course, there is also the option of manually adding the capability to the project
